src/dh/README.md

```diff
@@ -0,0 +1,72 @@
+# Diffie-Hellman Key Exchanges
+
+## What is Diffie-Hellman?
+
+The Diffie-Hellman key exchange protocol, named after Whitfield Diffie and Martin Hellman, is a protocol for two parties to create a shared secret over a public channel without revealing information about the key.
+
+The exchange is made possible by using commutativity in finite cyclic groups, particularly with ellitpic curve cyclic groups. That is to say, given a common generator point on an ellitpic curve over a finite field $G \in E(\mathbb{F}_p)$, two parties, alice and bob may choose secrets, $a \in \mathbb{F}_p$ and $b \in \mathbb{F}_p$ respectively, compute their respective elliptic curve points $[a]G = A \in E(\mathbb{F}_p)$ and $[b]G = B \in E(\mathbb{F}_p)$ (again respectively), and exchange these on a public channel. Finally, they may perform the elliptic curve point arithmetic on their received points with their secrets as follows.
+
+Alice:
+
+$$[a](B) = [a]([b]G) = [ab]G$$
+
+Bob:
+
+$$[b](A) = [b]([a]G) = [ba]G$$
+
+Finally:
+
+$$[ba]G = [ab]G$$
+
+As such, alice and bob have computed a shared secret $[ab]G \in E(\mathbb{F}_p)$ by sharing only $B$ and $A$ over a public channel, which, provided the ellitpic curve discrete log problem is intractable, leaks no useful information about $a, b \in \mathbb{F}_p$.
+
+This protocol is often used to exchange a cryptographic key to be used in a symmetric encryption algorithm and is used in protocols such as SSH, HTTPS, and a variant of it in the Signal Protocol.
+
+## What does a Diffie-Hellman Key Exchange look like?
+
+In practice, each of the two parties performs the following.
+
+1. Generate a local secret $a \in \mathbb{F}_p$ with a cryptographically secure pseudorandom number generator.
```

Comment:
I think it's time for ronk to have a PRNG

Reply:
true, i'd break it out into another pr first tho tbh. this pr is pretty limited in scope. might make sense to build separate abstractions which compose these things (ie key manager that does _all the things_)